### PR TITLE
MULE-9831: I cannot write an String to a file with an arbitrary encoding

### PR DESCRIPTION
**Pull request by marianogonzalez**

### CODE REVIEW

File: extensions/file/src/test/java/org/mule/extension/file/FileWriteTestCase.java [hunk 1]
```diff
@@ -133,6 +136,26 @@ public void writeStaticContent() throws Exception
         assertThat(content, is(HELLO_WORLD));
     }
 
+    @Test
+    public void writeWithCustomEncoding() throws Exception
+    {
+        final String defaultEncoding = muleContext.getConfiguration().getDefaultEncoding();
+        assertThat(defaultEncoding, is(notNullValue()));
+
+        final String customEncoding = Charset.availableCharsets().keySet().stream()
```
Comment (elrodro83) [inline extensions/file/src/test/java/org/mule/extension/file/FileWriteTestCase.java]: static import

File: extensions/file/src/test/java/org/mule/extension/file/FileWriteTestCase.java [hunk 2]
```diff
@@ -133,6 +136,26 @@ public void writeStaticContent() throws Exception
...
+                .filter(encoding -> !encoding.equals(defaultEncoding))
+                .findFirst()
+                .orElse(null);
+
+        assertThat(customEncoding, is(notNullValue()));
+        final String filename = "encoding.txt";
+
+        doWrite("write", filename, HELLO_WORLD, CREATE_NEW, false, customEncoding);
+        byte[] content = FileUtils.readFileToByteArray(new File(temporaryFolder.getRoot(), filename));
```
Comment (elrodro83) [inline extensions/file/src/test/java/org/mule/extension/file/FileWriteTestCase.java]: static import

File: extensions/file/src/test/java/org/mule/extension/file/FileWriteTestCase.java [hunk 3]
```diff
@@ -133,6 +136,26 @@ public void writeStaticContent() throws Exception
...
+
+        assertThat(Arrays.equals(content, HELLO_WORLD.getBytes(customEncoding)), is(true));
```
Comment (elrodro83) [inline extensions/file/src/test/java/org/mule/extension/file/FileWriteTestCase.java]: use an array matcher

File: extensions/ftp/src/test/java/org/mule/extension/ftp/FtpWriteTestCase.java [hunk 4]
```diff
@@ -142,6 +147,26 @@ public void writeStaticContent() throws Exception
...
+                .findFirst()
+                .orElse(null);
+
+        assertThat(customEncoding, is(notNullValue()));
+        final String filename = "encoding.txt";
+
+        doWrite("write", filename, HELLO_WORLD, CREATE_NEW, false, customEncoding);
+        InputStream content = (InputStream) readPath(Paths.get(testHarness.getWorkingDirectory()).resolve(filename).toString()).getPayload();
+
+        assertThat(Arrays.equals(IOUtils.toByteArray(content), HELLO_WORLD.getBytes(customEncoding)), is(true));
```
Comment (elrodro83) [inline extensions/ftp/src/test/java/org/mule/extension/ftp/FtpWriteTestCase.java]: same as FileWriteTestCase

File: extensions/ftp/src/test/java/org/mule/extension/ftp/FtpWriteTypeTestCase.java [hunk 5]
```diff
@@ -36,22 +36,15 @@
                 {"Ftp - String", new ClassicFtpTestHarness(), HELLO_WORLD, HELLO_WORLD},
                 {"Ftp - native byte", new ClassicFtpTestHarness(), "A".getBytes()[0], "A"},
                 {"Ftp - Object byte", new ClassicFtpTestHarness(), new Byte("A".getBytes()[0]), "A"},
-                {"Ftp - byte[]", new ClassicFtpTestHarness(), HELLO_WORLD.getBytes(), HELLO_WORLD},
```
Comment (elrodro83) [inline extensions/ftp/src/test/java/org/mule/extension/ftp/FtpWriteTypeTestCase.java]: why are these removed?

File: extensions/file/src/test/java/org/mule/extension/file/FileWriteTypeTestCase.java [hunk 6]
```diff
@@ -37,9 +37,6 @@
                 {"byte[]", HELLO_WORLD.getBytes(), HELLO_WORLD},
                 {"OutputHandler", new TestOutputHandler(), HELLO_WORLD},
                 {"InputStream", new ByteArrayInputStream(HELLO_WORLD.getBytes()), HELLO_WORLD},
-                {"Iterable", Arrays.asList(HELLO_WORLD, new HelloWorld()), HELLO_WORLD + HELLO_WORLD},
```
Comment (elrodro83) [inline extensions/file/src/test/java/org/mule/extension/file/FileWriteTypeTestCase.java]: why are these removed?

File: extensions/file/src/test/java/org/mule/extension/file/FileWriteTestCase.java (hunk 1 above)
Comment (marianogonzalez) [inline extensions/file/src/test/java/org/mule/extension/file/FileWriteTestCase.java]: done

File: extensions/file/src/test/java/org/mule/extension/file/FileWriteTestCase.java (hunk 2 above)
Comment (marianogonzalez) [inline extensions/file/src/test/java/org/mule/extension/file/FileWriteTestCase.java]: done

File: extensions/file/src/test/java/org/mule/extension/file/FileWriteTypeTestCase.java (hunk 6 above)
Comment (marianogonzalez) [inline extensions/file/src/test/java/org/mule/extension/file/FileWriteTypeTestCase.java]: we're also reducing the accepted types, thus these cases makes no sense

File: extensions/ftp/src/test/java/org/mule/extension/ftp/FtpWriteTestCase.java (hunk 4 above)
Comment (marianogonzalez) [inline extensions/ftp/src/test/java/org/mule/extension/ftp/FtpWriteTestCase.java]: yes, but in FTP. Although there's common ground, there're also important differences. The effort to reuse that code doesn't justify the effort. Already tried...

File: extensions/ftp/src/test/java/org/mule/extension/ftp/FtpWriteTypeTestCase.java (hunk 5 above)
Comment (marianogonzalez) [inline extensions/ftp/src/test/java/org/mule/extension/ftp/FtpWriteTypeTestCase.java]: same answer as before

File: extensions/ftp/src/test/java/org/mule/extension/ftp/FtpWriteTestCase.java (hunk 4 above)
Comment (elrodro83) [inline extensions/ftp/src/test/java/org/mule/extension/ftp/FtpWriteTestCase.java]: i meant same comments (static imports)

File: modules/file-extension-common/src/main/java/org/mule/runtime/module/extension/file/api/FileWriterVisitor.java [hunk 7]
```diff
@@ -28,25 +25,34 @@
...
 
     @Override
     public void visit(String content) throws Exception
     {
-        IOUtils.write(content, outputStream);
+        char[] payload = new char[content.length()];
+        content.getChars(0, content.length(), payload, 0);
+
+        IOUtils.write(payload, outputStream);
+        //try (OutputStreamWriter writer = new OutputStreamWriter(outputStream, encoding))
```
Comment (elrodro83) [inline modules/file-extension-common/src/main/java/org/mule/runtime/module/extension/file/api/FileWriterVisitor.java]: remove commented code
